src/main/java/org/tikv/txn/gc/GCWorker.java

```diff
@@ -0,0 +1,472 @@
+  private static final int gcMaxConcurrency = 128;
+  private final int gcWaitTime;// = 3 * 1000;
+  private final int gcWorkerLease;// = 2 * 1000;
+
+  public GCWorker(TiSession session) {
+    this.pdClient = session.getPDClient();
+    TiConfiguration conf = session.getConf();
+    this.gcDefaultRunInterval = conf.getGCRunInterval();
+    this.gcDefaultLifeTime = conf.getGCLifeTime();
+    this.gcDefaultConcurrency = conf.getGCConcurrency();
+    this.gcWaitTime = conf.getGCWaitTime();
+    this.gcWorkerLease = conf.getGCWorkerLease();
+    this.l_uuid = pdClient.grantLease(gcWorkerLease / 1000);
+    this.uuid = Long.toHexString(l_uuid);
+    logger.info("uuid = " + uuid + " " + l_uuid);
+    this.lastFinish = 0;
+    this.regionStoreClientBuilder = new RegionStoreClientBuilder(session.getConf(), session.getChannelFactory(), new RegionManager(session.getPDClient()));
+  }
+
+  @Override
+  public void close() throws Exception {
+    pdClient.unlock(GC_LEADER_UUID_KEY);
+    pdClient.revoke(l_uuid);
+    if (gcTaskThreadPool != null) {
+      gcTaskThreadPool.shutdown();
+    }
+    scheduler.shutdown();
+    gcWorkerScheduler.shutdown();
+  }
+
+  private TiTimestamp getTimestamp() {
+    return pdClient.getTimestamp(ConcreteBackOffer.newTsoBackOff());
+  }
+
+  private long getNow() { return getTimestamp().getPhysical(); }
+
+  // Note: Should not call start() more than once
+  public void start() {
+    gcWorkerScheduler.schedule(this::run, 0, TimeUnit.MILLISECONDS);
+  }
+
+  private void run() {
+    logger.info(String.format("[gc worker] %s starts", uuid));
+    scheduler.scheduleAtFixedRate(() -> {
+      try {
+        logger.debug("[gc worker] tick");
+        pdClient.keepLeaseAlive(l_uuid);
+        tick();
+      } catch (Exception e) {
+        logger.warn("[gc worker] gc fails to proceed", e);
+      }
+    }, 0, 1, TimeUnit.SECONDS);
+  }
+
+  private void tick() {
+    boolean isLeader = checkLeader();
+    if (isLeader) {
+      try {
+        leaderTick();
+      } catch (Exception e) {
+        logger.warn("[gc worker] check leader err: ", e);
+      }
+    }
+  }
+
+  private void leaderTick() {
+    if (gcIsRunning) {
+      logger.info(String.format("[gc worker] leader tick on %s: there's already a gc job running. skipped.", uuid));
+      return;
+    }
+
+    if (getNow() < lastFinish + gcWaitTime) {
+      gcIsRunning = false;
+      logger.info(String.format("[gc worker] leader tick on %s: another gc job has just finished. skipped.", uuid));
+      return;
+    }
+
+    long safePoint;
+    try {
+      safePoint = prepare();
+    } catch (Exception e) {
+      logger.warn("[gc worker] leader tick err: %s", e);
+      return;
+    }
+    if (safePoint == 0) {
+      gcIsRunning = false;
+      return;
+    }
+
+    gcIsRunning = true;
+    logger.info(String.format("[gc worker] %s starts the whole job, safePoint: %d", uuid, safePoint));
+    runGCJob(safePoint);
+    lastFinish = getNow();
+    gcIsRunning = false;
```

Comment:
It is async, and I believe we do not need to care about the `gcIsRunning` flag being unset because it only notifies `GCWorker` itself whether a job is running.

Reply:
If `gcIsRunning` is false, but the GCWorker is actually asynchronously processing GC tasks for each region, won't it went wrong?

Reply:
fixed